File: apps/worker/src/collectors/internal/indicators.ts
import type { Collector, CollectRequest, CollectResult, CollectorContext } from '../types';
import { fetchFIIIndicators } from '../../services/client';
import { sha256, nowIso } from '../../utils/date';
import { getFundIdByCode } from '../../pipeline/repo';

export const indicatorsCollector: Collector = {
  name: 'indicators',
  supports(request: CollectRequest) {
    return request.collector === 'indicators' && !!request.fund_code;
  },
  async collect(request: CollectRequest, _ctx: CollectorContext): Promise<CollectResult> {
    const code = String(request.fund_code || '').toUpperCase();
    const id = await getFundIdByCode(code);
    if (!id) throw new Error('FII_NOT_FOUND');

    const data = await fetchFIIIndicators(id);
    const dataJson = JSON.stringify(data);
    const dataHash = sha256(dataJson);
    const fetchedAt = nowIso();

    return {
      collector: 'indicators',
      fetched_at: fetchedAt,
      payload: {
        persist_request: {
          type: 'indicators',
          item: {
            fund_code: code,
            fetched_at: fetchedAt,
            data_json: data,
            data_hash: dataHash,
          },
        },
      },
    };
  },
};

File: go-worker/internal/collectors/indicators.go
package collectors

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/luizfelipeneves/api-fundo/go-worker/internal/db"
	"github.com/luizfelipeneves/api-fundo/go-worker/internal/httpclient"
	"github.com/luizfelipeneves/api-fundo/go-worker/internal/parsers"
)

// IndicatorsCollector collects fund indicators
type IndicatorsCollector struct {
	client *httpclient.Client
	db     *db.DB
}

// NewIndicatorsCollector creates a new indicators collector
func NewIndicatorsCollector(client *httpclient.Client, database *db.DB) *IndicatorsCollector {
	return &IndicatorsCollector{
		client: client,
		db:     database,
	}
}

// Name returns the collector name
func (c *IndicatorsCollector) Name() string {
	return "indicators"
}

// Collect fetches fund indicators
func (c *IndicatorsCollector) Collect(ctx context.Context, req CollectRequest) (*CollectResult, error) {
	code := parsers.NormalizeFundCode(req.FundCode)
	if verboseLogs() {
		log.Printf("[indicators] collecting indicators for %s\n", code)
	}

	// Get fund ID from database
	fundID, err := c.db.GetFundIDByCode(ctx, code)
	if err != nil {
		return nil, fmt.Errorf("failed to get fund ID: %w", err)
	}

	if fundID == "" {
		return nil, fmt.Errorf("FII_NOT_FOUND: %s", code)
	}

	// Fetch indicators from API
	url := fmt.Sprintf("%s/api/fii/historico-indicadores/%s/5", httpclient.BaseURL, fundID)

	var rawData interface{}
	err = c.client.GetJSON(ctx, url, &rawData)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch indicators: %w", err)
	}

	// Normalize indicators
	indicators := parsers.NormalizeIndicatorsAny(rawData)

	return &CollectResult{
		Data: IndicatorsData{
			FundCode: code,
			Data:     indicators,
		},
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}, nil
}

// IndicatorsData represents indicators data
type IndicatorsData struct {
	FundCode string
	Data     parsers.NormalizedIndicators
}
